### PR TITLE
Feature/search form

### PR DESCRIPTION
**Pull request by Kachyr**
#122 task 
I've add dropdown menu with search preferences and redirection to "Search-Results page" and moved styles in separate file.
TODO #37 (need to make some preferences with routing and query redirection)

### CODE REVIEW

File: src/components/SearchField.tsx [hunk 1]
```diff
@@ -1,38 +1,233 @@
...
+  'Romance',
+  'Mystery',
+  'Thriller',
+  'Animation',
+  'Action',
+  'Adventure',
+  'Crime',
+  'Fantasy',
+  'Science Fiction',
+
```
Comment (mrSelieznov) [inline src/components/SearchField.tsx:21]: Delete space

File: src/styles/SearchField.scss [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      height: 21px;
+
+      @media screen and(min-width: 992px) {
+        width: 28px;
+        height: 28px;
+      }
+    }
+
+
+    @media screen and(min-width: 992px) {
```
Comment (mrSelieznov) [inline src/styles/SearchField.scss:66]: Delete one apace

File: src/styles/SearchField.scss [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    .dropdown-wrapper{
+      display: grid;
+      grid-template-columns: 1fr;
+      grid-gap: 15px;
+      padding:15px;
+
+      @media screen and (min-width:992px) {
+        grid-template-columns:repeat(2, 1fr); ;
+
```
Comment (mrSelieznov) [inline src/styles/SearchField.scss:90]: space

File: src/styles/SearchField.scss [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        }
+
+        input {
+          font-size:16px;
+          text-align: center;
+          width:  60%;
+          height: 100%;
+        }
+        
+
```
Comment (mrSelieznov) [inline src/styles/SearchField.scss:134]: one space

File: src/styles/SearchField.scss (hunk 3 above)
Comment (Kachyr) [reply src/styles/SearchField.scss]: done

File: src/styles/SearchField.scss (hunk 4 above)
Comment (Kachyr) [reply src/styles/SearchField.scss]: done

File: src/components/SearchField.tsx (hunk 1 above)
Comment (Kachyr) [reply src/components/SearchField.tsx]: done

File: src/components/SearchField.tsx [hunk 5]
```diff
@@ -1,38 +1,232 @@
...
     setInputValue(event.target.value);
   };
 
-  const submitHandler = (event: React.SyntheticEvent) => {
-    event.preventDefault();
-    if (inputValue.trim()) {
-      console.log(inputValue);
-    }
-  };
+  const submitHandler = useCallback(
```
Comment (AnnaBolshak) [inline src/components/SearchField.tsx:59]: Could you give more specific names for all functions with names contain handler please?

File: src/components/SearchField.tsx [hunk 6]
```diff
@@ -1,38 +1,232 @@
...
+    (event: React.SyntheticEvent) => {
+      event.preventDefault();
+
+      // for proper using of query params need to refactor routing
```
Comment (AnnaBolshak) [inline src/components/SearchField.tsx:63]: should it be // TODO: point?

File: src/components/SearchField.tsx (hunk 5 above)
Comment (Kachyr) [reply src/components/SearchField.tsx]: I changed names for functions pls check if they good

File: src/components/SearchField.tsx (hunk 6 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:63]: its already in TODO #37 but not described, thank you)

File: src/components/SearchField.tsx [hunk 7]
```diff
@@ -1,38 +1,241 @@
...
         <input
+          onBlur={onBlurSearchInput}
           value={inputValue}
-          onChange={changeHandler}
+          onChange={searchInputChange}
           id="search"
           type="search"
           placeholder="Search"
           className="header-search"
+          onFocus={onFocusSearchInput}
```
Comment (AnnaBolshak) [inline src/components/SearchField.tsx:162]: Prefix "on" usually means event, the name of function is a verb

File: src/components/SearchField.tsx [hunk 8]
```diff
@@ -1,38 +1,241 @@
...
+                  checked={checkboxes.history}
+                />
+                <span>History</span>
+                <img
+                  className="check-img"
+                  src={checkboxes.history ? checkedImg : uncheckedImg}
+                  alt=""
+                />
+              </label>
+              <select
```
Comment (AnnaBolshak) [inline src/components/SearchField.tsx:222]: This select does not look like in the design, but maybe it is not crucial..
But in collapsed state its height is lower than other inputs:

image

And also without any label for input it is not clear what filter is set as None by default. The text for this default value could be something like "Select genre" or what ever you prefer, to give user understanding what criteria for search in this select

File: src/components/SearchField.tsx [hunk 9]
```diff
@@ -1,38 +1,241 @@
...
+import React, { useCallback, useState } from 'react';
+import { useHistory } from 'react-router';
+
+import '../styles/SearchField.scss';
 
 const loupeIcon = '/images/search-icon.svg';
 
+const genres = [
+  'none',
+  'Comedy',
```
Comment (AnnaBolshak) [inline src/components/SearchField.tsx:10]: It seems like it could be enum

**Review by AnnaBolshak — CHANGES_REQUESTED** [review]

File: src/components/SearchField.tsx [hunk 10]
```diff
@@ -1,38 +1,241 @@
...
+    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+      if (dateRange.fromDate >= dateRange.toDate) return;
+      setDateRange({ ...dateRange, fromDate: dateRange.fromDate + 1 });
+    },
+    [dateRange]
+  );
+
+  const fromDateInputDecrease = useCallback(
+    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+      if (dateRange.fromDate <= 1980) return;
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:108]: 1980 is a "magic number" - do not use such values in the middle of code - create const where name will describe value

File: src/components/SearchField.tsx [hunk 11]
```diff
@@ -1,38 +1,241 @@
...
+    },
+    [checkboxes]
+  );
+
+  const selectGenreHandler = useCallback(
+    (event: React.ChangeEvent<HTMLSelectElement>): void => {
+      setGenreToSearch(event.target.value);
+    },
+    []
+  );
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:149]: Cant get the point of callback usage

File: src/components/SearchField.tsx [hunk 12]
```diff
@@ -1,38 +1,241 @@
...
+      if (dateRange.toDate >= 2022) return;
+      setDateRange({ ...dateRange, toDate: dateRange.toDate + 1 });
+    },
+    [dateRange]
+  );
+
+  const toDateInputDecrease = useCallback(
+    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+      if (dateRange.toDate <= dateRange.fromDate) return;
+      setDateRange({ ...dateRange, toDate: dateRange.toDate - 1 });
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:125]: toDate-1.... See comment about renaming to toYear

File: src/components/SearchField.tsx [hunk 13]
```diff
@@ -1,38 +1,241 @@
...
+    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+      if (dateRange.fromDate <= 1980) return;
+      setDateRange({ ...dateRange, fromDate: dateRange.fromDate - 1 });
+    },
+    [dateRange]
+  );
+
+  const toDateInputIncrease = useCallback(
+    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+      if (dateRange.toDate >= 2022) return;
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:116]: magic number... 
why not new Date().getFullYear() ? so we dont need to redeploy code every year...

File: src/components/SearchField.tsx [hunk 14]
```diff
@@ -1,38 +1,241 @@
...
+  const mouseLeaveDropdown = useCallback(
+    (event: React.SyntheticEvent): void => {
+      setDropdown(false);
+    },
+    []
+  );
+
+  const fromDateInputIncrease = useCallback(
+    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+      if (dateRange.fromDate >= dateRange.toDate) return;
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:100]: hm, so do nothing on invalid date range?  How user will know that something is wrong?

File: src/components/SearchField.tsx [hunk 15]
```diff
@@ -1,38 +1,241 @@
...
+      setDateRange({ ...dateRange, fromDate: dateRange.fromDate + 1 });
+    },
+    [dateRange]
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:103]: what the point of useCallback here? on every change you reset dateRange = rerun callback function

File: src/components/SearchField.tsx [hunk 16]
```diff
@@ -1,38 +1,241 @@
...
 
-  const changeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
+  const [focus, setFocus] = useState<boolean>(false);
+
+  const [dropdown, setDropdown] = useState<boolean>(false);
+
+  const [dateRange, setDateRange] = useState<{
+    fromDate: number;
+    toDate: number;
+  }>({
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:40]: probably would be easier just to use separate states for dateFrom/dateTo

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]

File: src/components/SearchField.tsx [hunk 17]
```diff
@@ -1,38 +1,241 @@
...
+    fromDate: 1980,
+    toDate: 2021
```
[the remark this answers is not in the document]
Comment (Kachyr) [reply src/components/SearchField.tsx:42]: fixed

File: src/components/SearchField.tsx (hunk 10 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:108]: done

File: src/components/SearchField.tsx (hunk 13 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:116]: done

File: src/components/SearchField.tsx (hunk 12 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:125]: done

File: src/components/SearchField.tsx (hunk 16 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:40]: Now it works properly, may I leave it for now?

File: src/components/SearchField.tsx (hunk 11 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:149]: deleted it

File: src/components/SearchField.tsx (hunk 7 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:162]: replaced it with "whenFocusSearchInput"

File: src/components/SearchField.tsx (hunk 8 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:222]: fixed it

File: src/components/SearchField.tsx (hunk 9 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:10]: i dont know how to make my code working with enum, can you help me this please?

File: src/components/SearchField.tsx [hunk 18]
```diff
@@ -1,38 +1,241 @@
...
+        history.push(
+          `/search-results/title=${inputValue}?genre=${genreToSearch}?fromDate=${dateRange.fromDate}?toDate=${dateRange.toDate}?favorites=${checkboxes.favorites}?history=${checkboxes.history}`
+        );
+      }
+    },
+    [inputValue, history, checkboxes, genreToSearch, dateRange]
+  );
+
+  const onFocusSearchInput = useCallback(
+    (event: React.SyntheticEvent): void => {
```
[the remark this answers is not in the document]
Comment (Kachyr) [reply src/components/SearchField.tsx:74]: removed all "useCallback"

File: src/components/SearchField.tsx (hunk 14 above)
Comment (Kachyr) [reply src/components/SearchField.tsx:100]: I will add this functionality later

File: src/components/SearchField.tsx (hunk 9 above)
Comment (AnnaBolshak) [reply src/components/SearchField.tsx:9]: Could you explain a bit what is not working?

Comment (oleksandrstarov) [reply src/components/SearchField.tsx:9]: enum Genres {
  Drama='Drama',
...
}
use like this: Genres.Drama

File: src/components/SearchField.tsx [hunk 19]
```diff
@@ -1,38 +1,221 @@
...
+  const [focus, setFocus] = useState<boolean>(false);
+
+  const [dropdown, setDropdown] = useState<boolean>(false);
+
+  const [dateRange, setDateRange] = useState<{
+    fromYear: number;
+    toYear: number;
+  }>({
+    fromYear: 1980,
+    toYear: 2021
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:41]: avoid magic numbers, use const
toYear: 2021 >> new Date().getFullYear() - so no need to change this line every year 01.01

Comment (Kachyr) [reply src/components/SearchField.tsx:41]: replaced with const and new Date().getFullYear()

File: src/components/SearchField.tsx (hunk 14 above)
Comment (oleksandrstarov) [reply src/components/SearchField.tsx]: Please add TODO

File: src/components/SearchField.tsx (hunk 9 above)
Comment (oleksandrstarov) [reply src/components/SearchField.tsx:9]: @Kachyr any update on this?

Comment (Kachyr) [reply src/components/SearchField.tsx:9]: Done thank you)

File: src/components/SearchField.tsx (hunk 14 above)
Comment (Kachyr) [reply src/components/SearchField.tsx]: added #161

File: src/components/SearchField.tsx [hunk 20]
```diff
@@ -1,38 +1,224 @@
...
+            className="dropdown-filter"
+            onMouseEnter={mouseEnterDropdown}
+            onMouseLeave={mouseLeaveDropdown}>
+            <div className="dropdown-wrapper">
+              <div className="date-select-wrapper">
+                <span>
+                  <button onClick={fromDateInputDecrease}>-</button>
+                  <input
+                    type="number"
+                    min="1980"
```
Comment (oleksandrstarov) [inline src/components/SearchField.tsx:159]: >> lowerDateLimit

**Review by oleksandrstarov — APPROVED** [review]
Check comment

File: src/components/SearchField.tsx (hunk 20 above)
Comment (Kachyr) [reply src/components/SearchField.tsx]: done

**Review by oleksandrstarov — APPROVED** [review]

**Review by AnnaBolshak — APPROVED** [review]

**Pull request merged**